### PR TITLE
Issue 6747:Cache segment to host mapping

### PR DESCRIPTION
**Pull request by anju-c-das**
**Change log description**  
(2-3 concise points about the changes in this PR. When committing this PR, the committer is expected to copy the content of this section to the merge description box)

**Purpose of the change**  
(_e.g._, Fixes #666, Closes #1234)

**What the code does**  
(Detailed description of the code changes)

**How to verify it**  
(Optional: steps to verify that the changes are effective)

### CODE REVIEW

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java [hunk 1]
```diff
@@ -1282,13 +1301,52 @@ private StreamSegments getStreamSegments(final SegmentRanges ranges, long reques
...
+        final long requestId = requestIdGenerator.get(); // TODO: handle requestID and traceID
+        long traceId = LoggerHelpers.traceEnter(log, "updateStaleValueInCache", segmentName, errNodeUri, requestId);
+        CachedPravegaNodeUri cachedNodeUri = getSegmentEndpointFromCache(segmentId);
+        cachedNodeUri.getPravegaNodeUri().thenCombine(errNodeUri.getPravegaNodeUri(), (s1,s2) -> {
+            // in case if new request came for same segment, then it might have got updated since the entry is stale for this segment.
+                if(s1.getEndpoint().equals(s2.getEndpoint()) && s1.getPort() == s2.getPort()) {
+                    return getPravegaNodeUri(segmentName); // enforce cache refresh if stale value
+                }
+                else {
+                    endPointCacheMap.put(segmentId, new CachedPravegaNodeUri(new Timer(), cachedNodeUri.getPravegaNodeUri()));
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java:1330]: Same here

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java [hunk 2]
```diff
@@ -1282,13 +1301,52 @@ private StreamSegments getStreamSegments(final SegmentRanges ranges, long reques
...
+        Exceptions.checkNotNullOrEmpty(segmentName, "segmentName");
+        Segment segment = Segment.fromScopedName(segmentName);
+        long segmentId = segment.getSegmentId();
+        final long requestId = requestIdGenerator.get(); // TODO: handle requestID and traceID
+        long traceId = LoggerHelpers.traceEnter(log, "updateStaleValueInCache", segmentName, errNodeUri, requestId);
+        CachedPravegaNodeUri cachedNodeUri = getSegmentEndpointFromCache(segmentId);
+        cachedNodeUri.getPravegaNodeUri().thenCombine(errNodeUri.getPravegaNodeUri(), (s1,s2) -> {
+            // in case if new request came for same segment, then it might have got updated since the entry is stale for this segment.
+                if(s1.getEndpoint().equals(s2.getEndpoint()) && s1.getPort() == s2.getPort()) {
+                    return getPravegaNodeUri(segmentName); // enforce cache refresh if stale value
```
Comment (ShwethaSNayak) [inline client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java:1327]: Here you can update cache map with future object of nodeURI.

**Comment by ShwethaSNayak** [comment]
Please add unit tests for code coverage

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java [hunk 3]
```diff
@@ -226,6 +231,18 @@ public class ControllerImpl implements Controller {
...
     private final long timeoutMillis;
+
+    /**
+     * Maximum time that segment endpoint details are held in the cache.
+     */
+    private static final Duration SEGMENT_ENTRY_EXPIRATION_TIME = Duration.ofMinutes(5);
+    /**
+     * Maximum number of segment endpoint entries in the cache.
+     */
+    private static final int MAX_CACHE_SIZE = 1000000;
```
[the remark this answers is not in the document]
Comment (anju-c-das) [reply client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java]: Modified as suggested.

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java [hunk 4]
```diff
@@ -1301,11 +1359,18 @@ public CompletableFuture<PravegaNodeUri> getEndpointForSegment(final String qual
                 .whenComplete((x, e) -> {
                     if (e != null) {
                         log.warn(requestId, "getEndpointForSegment {} failed: ", qualifiedSegmentName, e);
+                    } else {
+                        //Populate cache with the segment endpointUri
+                        endPointCacheMap.put(segment.getSegmentId(), new CachedPravegaNodeUri(new Timer(), CompletableFuture.completedFuture(x)));
```
[the remark this answers is not in the document]
Comment (anju-c-das) [reply client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java]: Modified as suggested.

**Comment by ShwethaSNayak** [comment]
Add proper descriptions for below
**Change log description

Purpose of the change
What the code does

How to verify it**

**Comment by anju-c-das** [comment]
Closing as another new one will be raised.

**Pull request closed without merging**